python_http_client/async_client.py

```diff
@@ -0,0 +1,99 @@
+"""Asynchronous Python HTTP Client Module"""
+
+from collections import namedtuple
+import http
+from typing import Optional
+from urllib.error import URLError, HTTPError
+
+from aiohttp import ClientSession
+from aiohttp.client_exceptions import ClientConnectionError
+
+from .client import Client, Response
+from .exceptions import handle_error
+
+
+class AiohttpClientSessionError(ValueError):
+    """Invalid aiohttp.ClientSession error"""
+
+
+class AsyncResponse(Response):
+    def __init__(self, code, body, headers):
+        self._status_code = code
+        self._body = body
+        self._headers = headers
+
+
+class AsyncClient(Client):
+    """Main async python HTTP client class"""
+    def __init__(
+        self, *args, client_session: Optional[ClientSession] = None, **kwargs
+    ):
+        """Create async Python HTTP client instance
+
+        :param ClientSession client_session: aiohttp.ClientSession instance,
+                                             optional, defaults to None
+        """
+
+        super().__init__(*args, **kwargs)
+        self._aiohttp_client_session = client_session
+
+    @property
+    def client_session(self):
+        """aiohttp.ClientSession instance"""
+        return self._aiohttp_client_session
+
+    @client_session.setter
+    def client_session(self, session: ClientSession):
+        """Apply aiohttp.ClientSession instance
+
+        The only ClientSession instance can be applied. Otherwise
+        AiohttpClientSessionError will be raised.
+
+        :param ClientSession session: aiohttp.ClientSession instance
+        :raises AiohttpClientSessionError: internal ClientSession exists
+                                           or another external has already
+                                           been applied
+        """
+        if self.client_session:
```

Comment:
I would check here `_aiohttp_client_session` itself as our only condition is session existence. Any other logic that may be applied in `client_session` getter won't affect the decision.